domain/externalcontroller/state/state.go

```diff
@@ -127,57 +128,131 @@ func (st *State) UpdateExternalController(
+	ctx context.Context,
+	ci crossmodel.ControllerInfo,
+	modelUUIDs []string,
+	tx *sql.Tx,
+) error {
+	cID := ci.ControllerTag.Id()
+	q := `
 INSERT INTO external_controller (uuid, alias, ca_cert)
 VALUES (?, ?, ?)
   ON CONFLICT(uuid) DO UPDATE SET alias=excluded.alias, ca_cert=excluded.ca_cert`[1:]
 
-		if _, err := tx.ExecContext(ctx, q, cID, ci.Alias, ci.CACert); err != nil {
-			return errors.Trace(err)
-		}
+	if _, err := tx.ExecContext(ctx, q, cID, ci.Alias, ci.CACert); err != nil {
+		return errors.Trace(err)
+	}
 
-		addrsBinds, addrsAnyVals := database.SliceToPlaceholder(ci.Addrs)
-		q = fmt.Sprintf(`
+	addrsBinds, addrsAnyVals := database.SliceToPlaceholder(ci.Addrs)
+	q = fmt.Sprintf(`
 DELETE FROM external_controller_address
 WHERE  controller_uuid = ?
 AND    address NOT IN (%s)`[1:], addrsBinds)
 
-		args := append([]any{cID}, addrsAnyVals...)
-		if _, err := tx.ExecContext(ctx, q, args...); err != nil {
-			return errors.Trace(err)
-		}
+	args := append([]any{cID}, addrsAnyVals...)
+	if _, err := tx.ExecContext(ctx, q, args...); err != nil {
+		return errors.Trace(err)
+	}
 
-		for _, addr := range ci.Addrs {
-			q := `
+	for _, addr := range ci.Addrs {
+		q := `
 INSERT INTO external_controller_address (uuid, controller_uuid, address)
 VALUES (?, ?, ?)
   ON CONFLICT(controller_uuid, address) DO NOTHING`[1:]
 
-			if _, err := tx.ExecContext(ctx, q, utils.MustNewUUID().String(), cID, addr); err != nil {
-				return errors.Trace(err)
-			}
+		if _, err := tx.ExecContext(ctx, q, utils.MustNewUUID().String(), cID, addr); err != nil {
+			return errors.Trace(err)
 		}
+	}
 
-		// TODO (manadart 2023-05-13): Check current implementation and see if
-		// we need to delete models as we do for addresses, or whether this
-		// (additive) approach is what we have now.
-
-		for _, modelUUID := range modelUUIDs {
-			q := `
+	// TODO (manadart 2023-05-13): Check current implementation and see if
+	// we need to delete models as we do for addresses, or whether this
+	// (additive) approach is what we have now.
+	for _, modelUUID := range modelUUIDs {
+		q := `
 INSERT INTO external_model (uuid, controller_uuid)
 VALUES (?, ?)
   ON CONFLICT(uuid) DO UPDATE SET controller_uuid=excluded.controller_uuid`[1:]
 
-			if _, err := tx.ExecContext(ctx, q, modelUUID, cID); err != nil {
+		if _, err := tx.ExecContext(ctx, q, modelUUID, cID); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return nil
+}
+
+func (st *State) ModelsForController(
+	ctx context.Context,
+	controllerUUID string,
+) ([]string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	q := `
+SELECT 	uuid 
+FROM   	external_model 
+WHERE  	controller_uuid = ?`
+
+	var models []string
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		rows, err := tx.QueryContext(ctx, q, controllerUUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		for rows.Next() {
+			var row string
```

Comment:
Definitely rename this to `modelUUID` or `uuid` or some such.

Reply:
done